### PR TITLE
Only use actually mismatched elements for reporting in `torch.testing`

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57135 Only use actually mismatched elements for reporting in `torch.testing`**
* #57134 Use fixed number format for error messages in `torch.testing`
* #57133 Improve error messages for scalar comparisons in `torch.testing`

Currently all values are used for the reported absolute and relative differences. This usually works fine, but breaks down for the extremals:

```python
torch.testing.assert_close(torch.tensor([1.0, 0.0]), torch.tensor([2.0, 0.0]))
```

```
[...]
Greatest absolute difference: 1.0 at 0 (up to 1e-05 allowed)
Greatest relative difference: nan at 1 (up to 1.3e-06 allowed)
```

Although the second element is matching it is listed as offender for the greatest relative difference. The `NaN` stems from the `0 / 0` division.

To overcome this, we should only use the values that were considered a mismatch for the reported stats.

### CODE REVIEW

File: torch/testing/_asserts.py [hunk 1]
```diff
@@ -168,11 +168,14 @@ def _trace_mismatches(actual: Tensor, expected: Tensor, mismatches: Tensor) -> S
     dtype = torch.float64 if actual.dtype.is_floating_point else torch.int64
     a_flat = actual.flatten().to(dtype)
     b_flat = expected.flatten().to(dtype)
+    matches_flat = ~mismatches.flatten()
 
     abs_diff = torch.abs(a_flat - b_flat)
+    abs_diff[matches_flat] = 0
```
Comment (mruberry) [inline torch/testing/_asserts.py:174]: Add a comment here

File: torch/testing/_asserts.py [hunk 2]
```diff
@@ -168,11 +168,14 @@ def _trace_mismatches(actual: Tensor, expected: Tensor, mismatches: Tensor) -> S
...
     max_abs_diff, max_abs_diff_flat_idx = torch.max(abs_diff, 0)
 
     rel_diff = abs_diff / torch.abs(b_flat)
+    rel_diff[matches_flat] = 0
```
Comment (mruberry) [inline torch/testing/_asserts.py:178]: Also a comment here

**Review by mruberry — APPROVED** [review]
Nice fix; this should be separated from the stack since it's not a dependency; I suggest we add a comment to explain why the values are being set to zero

**Comment by pmeier** [comment]
Superseded by #57923.

**Pull request closed without merging**